datafusion/functions/src/unicode/left.rs

```diff
@@ -0,0 +1,245 @@
+#[derive(Debug)]
+pub(super) struct LeftFunc {
+    signature: Signature,
+}
+
+impl LeftFunc {
+    pub fn new() -> Self {
+        use DataType::*;
+        Self {
+            signature: Signature::one_of(
+                vec![Exact(vec![Utf8, Int64]), Exact(vec![LargeUtf8, Int64])],
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl ScalarUDFImpl for LeftFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "left"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        utf8_to_str_type(&arg_types[0], "left")
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        match args[0].data_type() {
+            DataType::Utf8 => make_scalar_function(left::<i32>, vec![])(args),
+            DataType::LargeUtf8 => make_scalar_function(left::<i64>, vec![])(args),
+            other => exec_err!("Unsupported data type {other:?} for function left"),
+        }
+    }
+}
+
+/// Returns first n characters in the string, or when n is negative, returns all but last |n| characters.
+/// left('abcde', 2) = 'ab'
+/// The implementation uses UTF-8 code points as characters
+pub fn left<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = as_generic_string_array::<T>(&args[0])?;
+    let n_array = as_int64_array(&args[1])?;
+    let result = string_array
+        .iter()
+        .zip(n_array.iter())
+        .map(|(string, n)| match (string, n) {
+            (Some(string), Some(n)) => match n.cmp(&0) {
+                Ordering::Less => {
+                    let len = string.chars().count() as i64;
+                    Some(if n.abs() < len {
+                        string.chars().take((len + n) as usize).collect::<String>()
+                    } else {
+                        "".to_string()
+                    })
+                }
+                Ordering::Equal => Some("".to_string()),
+                Ordering::Greater => {
+                    Some(string.chars().take(n as usize).collect::<String>())
+                }
+            },
+            _ => None,
+        })
+        .collect::<GenericStringArray<T>>();
+
+    Ok(Arc::new(result) as ArrayRef)
+}
+
+#[cfg(test)]
+mod tests {
+    use arrow::array::{Array, StringArray};
+    use arrow::datatypes::DataType::Utf8;
+
+    use datafusion_common::{Result, ScalarValue};
+    use datafusion_expr::{ColumnarValue, ScalarUDFImpl};
+
+    use crate::unicode::left::LeftFunc;
+    use crate::utils::test::test_function;
+
+    #[test]
+    fn test_functions() -> Result<()> {
+        #[cfg(feature = "unicode_expressions")]
+        test_function!(
+            LeftFunc::new(),
+            &[
+                ColumnarValue::Scalar(ScalarValue::Utf8(Some(String::from("abcde")))),
+                ColumnarValue::Scalar(ScalarValue::Int64(Some(2))),
```

Comment:
Totally not needed as this code is just moved, but I think you can write this more concisely with `from` if you wnat:



```suggestion
                ColumnarValue::Scalar(ScalarValue::from("abcde")),
                ColumnarValue::Scalar(ScalarValue::from(2u64)),
```

Reply:
(same thing applies to the rest of the tests in this file and and the others)